### PR TITLE
Connect roads automation

### PR DESCRIPTION
**Pull request by willjallen**
This feature has been discussed before (#4697) (#4981), but has apparently been abandoned. I'll try to take a crack at it. 

EDIT:
It's at a decent state now. I have done some playtesting in constrained environments, including testing the pathfinding on very large maps. Definitely needs playtesting in actual games.

**TODO**:
- [x] Connect road option should only show when road is unlocked
- [x] Workers still progress one tile at a time, even if they can skip ahead.
- [x] Workers should upgrade roads to railroads.
- [x] Workers should not build roads on jungles/forests if the civ has that property. The pathfinding already accounts for this.
- [x] Highlighted/valid tiles in should include owned, neutral AND friendly tiles.
- [ ] Consider impassable tiles that are made passable by a civ trait
- [x] HexMath.getDistance is not able to factor World-wrap in.
- [x] Heuristic should be between tile and destination, not its neighbor (depends on HexMath issue)
- [x] If a worker has enough MP, it should move and start building the road in one turn instead of moving, having 1 MP remaining and waiting until the next turn to start building.
- [x] Repair pillaged roads along path
- [x] Refractor ForceAutomateRoadConnection
- [x] Highlight possible path when clicking a tile

UI
- [x] There should be a notification when connect road is cancelled
- [x] There should be a notification when connect road is finished
- [x] All automatable units include the UI button "Connect road". This should be reserved for workers only.
- [x] Sound (SoundPlayer.playRepeated(UncivSound("metalhit"), 4))
- [x] Use the same UnitAction icon for the destination marker
- [ ] Different colors - potential destinations green to blue, selected - orangey
- [x] Show generated path when unit is selected
- [x] KeyboardBinding

**BUGS**:
- [x] The workers seem to stop automating the roads when using autoplay, but resume when going to the next turn manually.
- [x] When kicked off the path, a worker will start building a road on the tile it was kicked onto. Also, the worker will stay automated when it shouldn't.
- [x] I encountered crashes via Exception("The current tile of worker ${unit.label()} is not in the path to the destination tile") EDIT: I found one instance of this to be caused when another civ settles a city and kicks the worker onto an adjacent tile, which takes it off its path.
- [x] Crash on `unit.movement.canMoveTo(futureTile)`. I think I fixed this by using `canReachInCurrentTurn` instead of `canReach` but it's a hard bug to reproduce.

**OTHER**:
- [x] I am ignoring over-river penatlies for civ's that have not yet researched bridges. This technically produces suboptimal paths, but I figure this is okay since this extra cost might produce large deviations that will be made moot fairly early into the game.
- [ ] Some conversation in the discord about the cost function being strictly minimal w.r.t movement cost versus one that is less optimal and weighted towards a more direct connection.

### CODE REVIEW

File: core/src/com/unciv/logic/map/mapunit/MapUnit.kt [hunk 1]
```diff
@@ -381,6 +385,8 @@ class MapUnit : IsPartOfGameInfoSerialization {
     fun isMoving() = action?.startsWith("moveTo") == true
 
     fun isAutomated() = automated
+
+    fun isForceAutomatingRoadConnection() = action == UnitActionType.ForceAutomateRoadConnection.value
```
[the remark this answers is not in the document]
Comment (willjallen) [reply core/src/com/unciv/logic/map/mapunit/MapUnit.kt:389]: Wanted some distinction that this is a special case of an automation that has some user direction. 

Just `automatingRoadConnection` coincides with what already happens when a unit is automated, no indication that it's tied to a specific global state

**Comment by willjallen** [comment]
> https://thenounproject.com/icon/road-224428/ ? Looks almost perfect I'd say? Maybe a little too much detail. Need it converted?

Nice! That would help.

**Comment by willjallen** [comment]
image

It looks good!

**Comment by Lesiakower** [comment]
Good to see this pull request! Nevertheless, I have one question:
Will this functionality take Iroquois' unique (which makes forest and jungle tiles natural roads) under account?

**Comment by willjallen** [comment]
The plan is to conduct a shortest path search between the start and end tile, where we assume that every tile visited already has a road connection. That way, the path we get out of this search will be the shortest movement mediated by road path and should in turn tell us where to place the road.

We will save this path as a list of vectors to the assigned unit and have them 
- build along it if possible
- skip and move to the next if a road is already present
- repair (road only) if pillaged

If the unit is blocked or anything prevents it from progressing it will just cancel, clear all the data, and return to a non-automated state.

> Will this functionality take Iroquois' unique (which makes forest and jungle tiles natural roads) under account?

Good point. I suppose I can add a special case for the Iroquois' to not build roads on forest and jungle tiles after it retrieves the path, however I believe this benefit does not extend to railroads and so that needs to be turned off when that is unlocked. Hopefully someone will correct me if I'm wrong about that.

@yairm210 You seem to be the last person to think about this so I'd like to hear your thoughts on this outline.

**Comment by willjallen** [comment]
>  conduct a shortest path search between the start and end tile, where we assume that every tile visited already has a road connection. That way, the path we get out of this search will be the shortest path mediated by road and should in turn tell us where to place the road.

I'm beginning to think the search solution might be relatively easy. Looking at the movement cost code from `getMovementCostBetweenAdjacentTiles` we have:
```kotlin
if (from.getUnpillagedRoad() == RoadStatus.Railroad && to.getUnpillagedRoad() == RoadStatus.Railroad)
    return RoadStatus.Railroad.movement + extraCost
```
and
```kotlin
if (areConnectedByRoad && (!areConnectedByRiver || civ.tech.roadsConnectAcrossRivers))
    return unit.civ.tech.movementSpeedOnRoads + extraCost
```
which implies that, if we assume every tile has a road connection on the search, then the movement cost is constant for all tiles. This means a simple BFS is guaranteed to return the ideal road location.

**Comment by willjallen** [comment]
https://github.com/yairm210/Unciv/assets/30008534/622a6dc5-2ec4-494a-8824-f11bbef14f3a

I have it in a semi-working state. Definitely needs some game testing, but the framework is there. One thing to notice is at the end of the video, the worker opts build parallel roads next to existing ones. This is could be solved with an A* algorithm, which would give negative weight to tiles with roads already on them. I'll consider doing this, it would also improve the road pathing between cities for automated units. It could also give a nice framework for more robust AI strategies generally.

I also plan to have the worker upgrade roads to railroads along the path, if available. 

Lastly, the worker will progress one tile at a time towards the destination, regardless of how many existing roads there are between. For example, starting at position A with destination B separated by 10 tiles w/ road and the last tile without a road would have the worker crawl each tile one at a time instead of just jumping to the last tile.

**Comment by AdityaMH** [comment]
add a condition "ignore "certain" tile or call it as road for "certain" tile if player or ai meet Iroquois ability or road things.

**Comment by willjallen** [comment]
> Thus a Worker on a Mountain... 

This brings up a good point that my check for "valid" tiles to search and build roads on is
```kotlin
{tile: Tile -> tile.isLand && !tile.isImpassible() && unit.civ.hasExplored(tile) && (tile.getOwner() == unit.civ || tile.getOwner() == null)}
```
So I should consider impassable tiles that are made passable by a civ trait, as well as foreign civ tiles we are allowed to build in (is there a check for this somewhere?)

---

> Use the same UnitAction icon for the destination marker?

> Different colors - potential destinations green to blue, selected - orangey?

> better a KeyboardBinding, goes right after the sound in onActivation.

> HexMath.getDistance is not able to factor World-wrap in.

> still would call it AutomateRoadConnection not ForceAutomateRoadConnection

> SoundPlayer.playRepeated(UncivSound("metalhit"), 4) maybe...

👍 

> Less java semicolons.

I've spent a lot of time in C++ recently so muscle memory is tough.

---

> Left side too cluttered

Might consider allowing that whole menu to be hidden. Strip the labels of the buttons, make them smaller and putting them inside the unit box. Just a row of icons. Then have an arrow ^ that would extend upwards to reveal what we already have now. Users can keep it on if they prefer the verbose descriptions for each action, but I suspect once the names of the buttons are learned they'll want more screen real estate instead.

---

> because automateUnitMoves will give precedence to tryHealUnit

This feels fine to me. I have yet to add checks for moving to tiles where we would be captured as a cancel condition, but I will. I want the unit to be overly sensitive to perturbations since it's not as smart as a fully automated unit.

> Or did it not like there was a parked General in its way?

Yup. This is part of the "moves just one tile at a time when it could go further" issue. Existing roads along the path don't really matter. The unit will cancel if it tries to go to the next tile in the path and can't
```kotlin
currTileIndex < pathToDest.size - 1 -> { // Try to move to the next tile in the path
                        val nextTile = unit.civ.gameInfo.tileMap[pathToDest[currTileIndex + 1]]
                        if(unit.movement.canMoveTo(nextTile) && unit.movement.canReach(nextTile)){
                            unit.movement.moveToTile(nextTile)
                            return
                        }else{ // Worker can't move to the next tile.
                            stopAndCleanAutomation()
                        }
                    }
```

---

> Then you can write the Developer wiki entry on how to set that up for Unciv

Android studio is based on idea, so it's almost identical.

**Comment by willjallen** [comment]
image

> Some conversation in the discord about the cost function being strictly minimal w.r.t movement cost versus one that is less optimal and weighted towards a more direct connection.

It is probably worth revisiting this. In the picture above, when attempting to connect to the road on the left, it auto cancels because the minimal movement cost path is along the existing roads. However, players may want to force this direct connection.

One solution is to add a static penalty for tiles further then a radius x around the destination tile would work, where x is tunable. This way, roads inside the radius *can* be reused and roads outside of it would not.

**Comment by willjallen** [comment]
> resulted in it automating to chop down a Jungle on a Luxury entirely elsewhere

hmm... this shouldn't happen. Only three things take precedence over the road building - healing, running away and `shouldClearTileForAddInCapitalUnits`. As soon as the worker goes off its path, `automated` gets set to false. Is it possible the Jungle was somewhere along the path?

**Comment by willjallen** [comment]
Fair enough. I am unable to reproduce this level of deviation on a fresh save with constant perturbations by nearby barbarians, so i'll write this one off as high energy space particles flipping bits in your machine.

**Comment by willjallen** [comment]
> Also, I advocate documenting every single `@Suppress` and `!!`...

My paradigm for using `!!` has been either "I want this to crash if it's null" under the premise that it being null (in context) is a bug, not a feature, and better to crash fast than produce cryptic undesired behavior later -- or "there is an explicit/obvious null check right above this". I'll document this explicitly.

> Without, meaning branch freshly pulled and rebased onto 4.9.5, it crashes instead. So something makes my Workers receive the ConnectRoad marker but not the automatedRoadConnectionPath...

I hadn't been actively thinking about interoperability or testing between game versions, so i'll keep that in mind going forward. 

Is this a save with a game state originating from an earlier commit inside this PR, or one from 4.9.5 release? There's no expectation of stability on the former. Some light testing by creating a save game in 4.9.5, building some roads, then rebasing this PR onto 4.9.5 and loading the game produces no obvious issues or crashes.

> instead of that filterIndexed - think about it.

Yup. No shortcut for learning the idioms of a new language!

> Too much code distributed over other files. Maybe maximize locality, put as much as possible in one new object file dealing exclusively with this feature.

There isn't much to localize, aside from moving the `automateConnectRoad` function. The extent of the reach beyond necessity (keyboard bindings & ui stuff) is storing the destination and path in the `MapUnit` class. 

If there is any place that would benefit 

>  But display the state "I'm supposed to build a road to, but just got the order - or got off track and had to forget my path" as one single-tile overlay...

👍 

> The HexMath.getDistance worldWrap problem

First I should mention I set up the initial heuristic wrong. It should be between the current tile to the goal tile, not neighbors (for this particular Euclidean distance heuristic). I.e the `to: Tile` parameter is ignored. 

In any case, I think I will just leave it empty and rely on Dijkstra. I have not encountered a situation in which the game hangs even a little because the search takes too long, even on massive maps. (However, the exponential nature of expanding search spaces could make a 1-tile difference in continent size the difference, but i'm happy to wait for a PR that will never come). 

All that we would get from implementing a heuristic is risking a case in which distance is not admissible and messing up the search (I don't think this can ever be the case, even on a complex environment such as Civ, but i'd rather keep that monster in the closet)

> "Log.debug" 

Yeah, consistency is good.

**Comment by willjallen** [comment]
So this is the state we're in now. I think decently close to a pre-release version. I only have so much time to do full game run-throughs, so will need fuzzing-by-the-masses. The remaining bug is pretty critical imo.

I'd appreciate some guidance on "owned, neutral AND friendly tiles." as well as the "impassable tiles that are made passable by a civ trait". Tracking down uniques is a real chore. 

**TODO**:
- [ ] Highlighted/valid tiles in should include owned, neutral AND friendly tiles.
- [ ] Consider impassable tiles that are made passable by a civ trait
- [ ] Repair pillaged roads along path


**UI**

- [ ] Different colors - potential destinations green to blue, selected - orangey


**BUGS**:
- [ ] The workers seem to stop automating the roads when using autoplay, but resume when going to the next turn manually.


**OTHER**:
- [ ] Some conversation in the discord about the cost function being strictly minimal w.r.t movement cost versus one that is less optimal and weighted towards a more direct connection.

**Comment by willjallen** [comment]
> There isn't much to localize, aside from moving the automateConnectRoad function. The extent of the reach beyond necessity (keyboard bindings & ui stuff) is storing the destination and path in the MapUnit class.

I reconsidered this and refactored finding the road path into a new MapPathing class. This made sense since I want to show previews of the path by highlighting tiles in the UI. 

As an aside, this also gives a nice framework to eventually rework the monstrosity of the UnitMovement functions into search calls. 

>  I regularly dream of writing one that would finally solve all the detours units are taking on any Archipelago map, with intelligent threat, tile damage, stopover placement, and so forth, evaluation - with provisions for stepwise refinement as Amit describes for speed but adapted for turns changing the map, modifiable by situation with some lambdas.

And a place for this.

**Comment by willjallen** [comment]
> Exactly - almost. I believe your class would still need to help differentiating stopovers/pass-throughs for multi-turn movement

I think you would just need to co-opt the tile.neighbor function with a state transition function. Requires some thinking, for another day.

> Agreed. I have a little nagging feeling we might get some surprises when >1 Workers are ordered to build the same road, either to coop from the same end (which manually often works really well), or from both ends - but I'd expect that to be relatively benign.

So far i've had no issues with this. However, I think we may need to disable the AI `tryConnectCities` functionality for human players for now, since it still uses BFS and only considers cities connected within that function. Leaving it on could result in a situation where the AI starts constructing a parallel, less optimal road, to one that you've already built.

I'm in discussion with another dev reworking the worker automation to eventually rework `tryConnectCities` into building a MST weighted sub-graph, which will fix this and allow the AI to construct more cost&movement efficient city connections.

**Comment by willjallen** [comment]
I'm going to mark this as ready for review. 

I have spent a lot of time testing in toy maps and played one full game all the way through. I have not experienced any crashes since the first few commits in this PR, and no game breaking bugs. I fully expect there to be some, but I doubt I will be able to find them with the time I have.

**Comment by yairm210** [comment]
First off - this looks good, definitely better than the nothing we have now: D

Regarding pathing, step 1 should be *building edge cases as tests*, NOT as external jsons! We've discussed several cases in the past but didn't document them, this should be a productive step even if no actual change comes off it in the end

**Review by yairm210 — APPROVED** [review]

**Comment by yairm210** [comment]
Simple example: field of plains, unit with 1 movement, tiles ABCD where B,C are hills
Sum of movement in straight line is 5 (2 hills plus plains), sum of going around hills is 4 (4 plains), BUT fastest route is through the hills (3 turns vs 4)
Naive Dijkstra would fail this test

**Comment by tuvus** [comment]
I think we should put the connect roads button under "show more". Also the button isn't working for me in Android Studio and on mobile.

**Comment by Lesiakower** [comment]
> Also the button isn't working for me in Android Studio and on mobile.

For me on Desktop (Linux) too...

**Comment by willjallen** [comment]
I'm unable to reproduce this on windows desktop. Is it just the button? Does the hotkey 'c' work?

**Comment by Lesiakower** [comment]
> Which button, the "Connect road" on the left by the other unit actions, or the icon button that appears when you tap a destination tile? Does it work if you long press or right click? Does the hotkey 'c' work?

For me, it is the button on the left that doesn't work. If I long press or right click it nothing happens. If I tap 'c' it works, in a sense that the "Connect road" button is activated. But the problem is that this button ITSELF doesn't work properly. At least for now, it's like a second "next unit" button. If I click it (with 'c' or with right-click) the game skips the worker and proceeds to the next unit.

**Comment by Lesiakower** [comment]
> Maybe one of the settings? Edit: Yup, it's "Auto Unit Cycle".

But should it work like that? Playing without "Auto Unit Cycle" will be bismal...

**Comment by willjallen** [comment]
@tuvus @Lesiakower Found it, easy fix. Thanks for the report!

**Comment by touhidurrr** [comment]
Hi @willjallen ! Thanks for your effort to pull this off ! I opened #4981 but forgot about it myself. It is good to see that someone has picked it up and completed it !
I might be late to thank you but better late than never. So, thanks for your work!
